### ISSUE TITLE
Authorized routes

### ISSUE THREAD

**Issue by steida** [4Catalyzer/found#73]
I'm not sure how to implement authorised routes. I need to check Redux state.users.viewer to enforce redirect with requested location. Does it belong to routeConfig or it should be a higher order component? Can we use `<Redirect />` component in a view?

**Comment by taion** [comment]
It depends how you want to handle it.

If you want to do a dynamic redirect based on what's in the store, I would put `store` on `matchContext`, then `throw RedirectException` in `render` on some route.

Personally, we don't redirect users on auth errors – we just render a login prompt instead of the page. We handle this by doing `throw HttpError(401)`, then using a `renderError` method that looks like:

```js
  renderError({ error }) { // eslint-disable-line react/prop-types
    switch (error.status) {
      case 404: return <NotFoundPage />;
      case 401: return <LoginPage />;
      default: return null;
    }
  },
```

**Comment by steida** [comment]
I would like to save previously request route to redirect after successful auth.

**Comment by taion** [comment]
Sure, so either do:

```js
throw new RedirectException({
    pathname: '/login',
    state: { previousLocation: location },
})
```

Or if you use the error handler approach, you get that for free (just do `router.replace(location)` afterward).

**Comment by steida** [comment]
Thank you.

**Comment by steida** [comment]
Btw, you can throw real JS errors via https://github.com/JsCommunity/make-error

**Comment by steida** [comment]
@taion throw HttpError is wrong btw, `throw new HttpError` https://github.com/4Catalyzer/found/issues/74

**Comment by taion** [comment]
Oops.

`HttpError` and `RedirectException` aren't real JS errors because they're used just for signalling; I don't want to incur the cost of messing with stack traces there.

**Comment by steida** [comment]
Btw, what is the recommended Route prop for requireAuth routes? getData? render?

**Comment by taion** [comment]
Use `getData` (and possibly return a promise) if you need to fetch additional data to figure out whether to block authorization.

Throw `HttpError` or `RedirectException` in `render`.

**Comment by steida** [comment]
Thank you. Full-fledged demo in soon in github.com/este/este.

**Comment by steida** [comment]
> Throw HttpError or RedirectException in render.

Hmm, but I don't have access to Redux store in render, only in getData.
I need to check `state.users.viewer`.

**Comment by taion** [comment]
Use `matchContext` per https://github.com/4Catalyzer/found#data-or-getdata

Then you can access it on `match.context` in `render`.

**Comment by taion** [comment]
Oh, I guess I didn't have that in the docs.

**Comment by steida** [comment]
I have `matchContext={{ store }}` on ConnectedRouter.

**Comment by taion** [comment]
Great, so the object you get as the argument to `render` has a `match` property, and you can use `match.context.store` in `render`.

**Comment by steida** [comment]
With flowtype, I could easily navigate the source code to read that, but it would not solve the issue I haven't vacation for years anyway. Thank you.

**Comment by steida** [comment]
I believe this should work:

`const RequireViewerRoute = props => <Route {...props} />;`

TypeError: type.createRoute is not a function

How do you create require auth routes?

**Comment by steida** [comment]
This seems to clunky

```jsx
// @flow
import HttpError from 'found/lib/HttpError';
import React from 'react';
import { makeRouteConfig, Route } from 'found/lib/jsx';

// Pages
import App from './App';
import FieldsPage from '../fields/FieldsPage';
import HomePage from '../home/HomePage';
import IntlPage from '../intl/IntlPage';
import MePage from '../me/MePage';
import OfflinePage from '../offline/OfflinePage';
import SignInPage from '../auth/SignInPage';
import TodosPage from '../todos/TodosPage';
import UsersPage from '../users/UsersPage';

const requireViewer = {
  getData({ context: { store } }) {
    if (!store.getState().users.viewer) {
      throw new HttpError(401);
    }
  },
};

const routeConfig = makeRouteConfig(
  <Route path="/" Component={App}>
    <Route Component={HomePage} />
    <Route path="fields" Component={FieldsPage} />
    <Route path="intl" Component={IntlPage} />
    <Route {...requireViewer} path="me" Component={MePage} />
    <Route path="offline" Component={OfflinePage} />
    <Route path="signin" Component={SignInPage} />
    <Route path="todos" Component={TodosPage} />
    <Route path="users" Component={UsersPage} />
  </Route>,
);

export default routeConfig;
```

**Comment by taion** [comment]
The current API looks a little bit like https://github.com/4Catalyzer/found/blob/master/src/jsx/Redirect.js

So you add a static `createRoute` on the component that returns an object (which can be of a class) that has the relevant properties.

We have something internally that looks like:

```js
function createRoute(props) {
  return {
    ...props,
    render: myRender,
  };
}

function Route() {
  invariant(
    false,
    '<Route> is only for configuration and should not be rendered.',
  );
}

Route.createRoute = createRoute;
```

Note that the `render` hook _always_ works on navigation, so unlike with React Router's `onEnter`, you can just nest all of your routes that require auth under a single (potentially pathless) parent route that checks for auth in `render`.

Also, I'm planning on eventually merging https://github.com/4Catalyzer/found/pull/68, which will sidestep the whole `createRoute` business with defining custom route classes.

**Comment by steida** [comment]
Sorry, I don't understand your explanation.

"So you add a static createRoute on the component"
Which component? Route component? View component?

"that returns an object"
Which object? Why?

"(which can be of a class"
What else can it be? Function? String? Whatever?

"that has the relevant properties."
What are relevant properties?

Sorry, but I am probably missing something.

**Comment by taion** [comment]
```js
function AuthorizedRoute() {}

AuthorizedRoute.createRoute = props => ({
  render: checkAuthorized,
  ...props,
});
```

Then use this `<AuthorizedRoute>` in your route config.

**Comment by steida** [comment]
Full example of Found router with SSR and hot reloading: https://github.com/este/este

Thank you @taion 😍

**Comment by hollandThomas** [comment]
This does not seem to work for me

```javascript
const routeConfig = makeRouteConfig(
  <Route path="/" Component={Page}>
    <Route Component={LandingPage} />
    <AuthorizedRoute path="test" Component={Test} />
    <Route path="login" Component={Login} />
    <Route path="register" Component={Register} />
  </Route>,
);
```

with

```javascript
const AuthorizedRoute = () => {};
AuthorizedRoute.createRoute = props => ({
  ...props,
  render: ({ Component, matchContext, props }) => {
    const state = matchContext.store.getState();
    if (!state.get('auth').get('isAuthorized')) throw new HttpError(401);
    return <Component {...props} />;
  },
});
```

which is inspired by [this](https://github.com/este/este/blob/4b226059771367edf4828eb7c98fc6b1ea8f3063/src/browser/app/routeConfig.js) (using found@0.2.1). I believe 'match.context' has since then been renamed to 'matchContext' or at least so say the react developer tools. Either way, both do not work.

When I navigate to `/test`, there is a blank white page and nothing more.

I see that you merged #68 by now and suspect this is the reason this is not working anymore(?) - if so, could you provide insights on how to alter the code? Thank you!

**Comment by taion** [comment]
You just do like:

```js
import Route from 'found/lib/Route';

export default class AuthorizedRoute extends Route{
  render({ Component, matchContext, props }) {
    // ...
  }
}
```

Basically just like https://github.com/4Catalyzer/found/blob/52917b827a6756c0e80d8c38a007abe2ebe67d93/src/Redirect.js.

**Comment by hollandThomas** [comment]
I'm sorry but I still seem to be missing something. This will render whatever Component is passed into it but it seems this code just gets ignored because the `console.log(state)` is not executed.

```javascript
export default class AuthorizedRoute extends Route {
  render({ Component, matchContext, props }) {
    const state = matchContext.store.getState();
    console.log(state);
    if (!state.get('auth').get('isAuthorized')) throw new HttpError(401);
    return <Component {...props} />;
  }
}
```

**Comment by taion** [comment]
Hmm, that code should throw, actually. It should be `match.context`.

**Comment by hollandThomas** [comment]
Are you sure it's not `matchContext`? Maybe I'm misinterpreting the React Developer Tools.

<img width="811" alt="bildschirmfoto 2017-10-26 um 00 37 30" src="https://user-images.githubusercontent.com/20690465/32026863-468c1f5c-b9e6-11e7-98dd-70b20395d66b.png">


I mean, for me, it does not make a difference. Even something like this just renders the component and not the `<div>test</div>`. That's what I mean by "this code just gets ignored.

```javascript
export default class AuthorizedRoute extends Route {
  render({ Component, match, props }) {
    const state = match.context.store.getState();
    console.log('state');
    if (!state.get('auth').get('isAuthorized')) throw new HttpError(401);
    return <div>test</div>;
  }
}
```

**Comment by taion** [comment]
That's really odd. Redirects use exactly this mechanism, and I know they work.

Can you check the output of `makeRouteConfig` and verify that you do actually have an `AuthorizedRoute` object in the place of your authorized route, rather than the `Route` object for the other routes?

**Comment by hollandThomas** [comment]
This is the output. There is in fact an authorized route.

<img width="814" alt="bildschirmfoto 2017-10-26 um 00 55 04" src="https://user-images.githubusercontent.com/20690465/32027258-5e2acb98-b9e8-11e7-9707-c7f9ea3aa354.png">

**Comment by taion** [comment]
Okay, and if you take that object, is its `render` property what you expect?

Also, if you just define that `render` method on the route itself, does it still not work?

**Comment by hollandThomas** [comment]
Alright this works for me. Still no clue though why the custom AuthorizedRoute does not work.

```javascript
const requireAuth = Component => ({
  render: ({ match, props }) => {
    const isAuthenticated = match.context.store
      .getState()
      .get('auth')
      .get('isAuthenticated');
    if (!isAuthenticated) return <Login {...props} />;
    return <Component {...props} />;
  },
});

const routeConfig = makeRouteConfig(
  <Route path="/" Component={Page}>
    <Route Component={LandingPage} />
    <Route {...requireAuth(Register)} path="test" />
    <Route path="chat" Component={ChatPage} query={ChatPageQuery} />
    <Route path="dashboard" Component={Dashboard} />
    <Route path="login" Component={Login} />
    <Route path="register" Component={Register} />
  </Route>,
);
```

Thank you for working through this with me! :) Let me know if I can be of any help if you want to investigate this further at some point.

**Comment by taion** [comment]
It's odd to me that this would happen. For a sanity check, could you see if `<Redirect>`s work in your app? They go through literally the same logic, so it'd be really confusing if your own custom route class did not work, but `<Redirect>` did.

**Comment by taion** [comment]
^ The above was a weird React Hot Loader interaction that will be fixed in https://github.com/4Catalyzer/found/pull/138.

**Comment by taion** [comment]
@hollandThomas Actually, can you confirm that you were using RHL?

**Comment by hollandThomas** [comment]
Yes, I was. react-hot-loader@3.1.1
